dali/python/nvidia/dali/pipeline.py

```diff
@@ -325,6 +326,18 @@ def __init__(self,
         elif output_ndim is not None and output_ndim < 0:
             raise ValueError(f"`output_ndim` must be non-negative. Found value: {output_ndim}.")
         self._output_ndim = output_ndim
+        Pipeline._pipes.add(self)
+
+    _pipes = set()  # this is necessary for clean exit
+
+    def __del__(self):
+        self._shutdown()
+
+    def _shutdown(self):
+        if self._pipe:
+            del self._pipe      # delete the backend pipeline, shutting it down
+            self._pipe = None
```

Comment:
Do we need both of these lines or just one?

Reply:
We certainly need the second one, or the field would disappear and references to it would cause missing attribute error. I'm not sure if this `del` does anything special, really. It definitely sends a message to the reader if this code.

Reply:
Sure

Reply:
Actually, now that I think of it, we should probably expose some kind of shutdown function from the backend and don't rely at all on `del` or assigning `None` doing its job.